### PR TITLE
API to create arrays from pure streams and string literals

### PR DESCRIPTION
**Pull request by rnjtranjan**

### CODE REVIEW

**Comment by rnjtranjan** [comment]
Fixes: https://github.com/composewell/streamly/issues/2385

File: core/src/Streamly/Internal/Data/Array/Mut.hs [hunk 1]
```diff
@@ -84,3 +91,10 @@ getSlicesFromLen from len =
 fromStream :: (MonadIO m, Unbox a) => Stream m a -> m (MutArray a)
 fromStream = fromStreamD
 -- fromStream (Stream m) = P.fold write m
+
+fromByteStr# :: Addr# -> Identity (MutArray Word8)
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Array/Mut.hs:95]: Why `Identity`?

File: core/src/Streamly/Internal/Data/Array/Mut/Type.hs [hunk 2]
```diff
@@ -1908,6 +1912,12 @@ writeWith elemCount =
 write :: forall m a. (MonadIO m, Unbox a) => Fold m a (MutArray a)
 write = writeWith (allocBytesToElemCount (undefined :: a) arrayChunkBytes)
 
+writePure :: Unbox a => Fold Identity a (MutArray a)
+writePure = FL.morphInner (Identity . unsafePerformIO) write
+
+fromPure :: Unbox a => Stream Identity a -> Identity (MutArray a)
+fromPure = D.fold writePure
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Array/Mut/Type.hs:1919]: Is this safe to do?

File: core/src/Streamly/Internal/Data/Array/Generic.hs [hunk 3]
```diff
@@ -113,6 +118,15 @@ writeWith elemCount = unsafeFreeze <$> MArray.writeWith elemCount
 write :: MonadIO m => Fold m a (Array a)
 write = fmap unsafeFreeze MArray.write
 
+fromPureStream :: Stream Identity a -> Array a
+fromPureStream x =
+    unsafePerformIO $ fmap (unsafeFreeze) (MArray.fromPureStream x)
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Array/Generic.hs:123]: Why can't you just use `MArray.fromStream` here instead?

File: core/src/Streamly/Internal/Data/Array/Type.hs [hunk 4]
```diff
@@ -495,6 +499,60 @@ writeWith elemCount = unsafeFreeze <$> MA.writeWith elemCount
...
+-- expression in a particular case?
+--
+-- For this reason array creation folds have a MonadIO constraint. Pure folds
+-- could be unsafe and dangerous. This is dangerous especially when used with
+-- foldMany like operations.
+--
+-- >>> unsafePureWrite = Array.unsafeMakePure Array.write
+--
+{-# INLINE unsafeMakePure #-}
+unsafeMakePure :: Monad m => Fold IO a b -> Fold m a b
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Array/Type.hs:525]: We shouldn't expose this.

Comment (adithyaov) [reply core/src/Streamly/Internal/Data/Array/Type.hs:525]: This can be misused.

File: core/src/Streamly/Internal/Data/Stream/StreamD/Generate.hs [hunk 5]
```diff
@@ -1143,35 +1144,35 @@ fromFoldableM = Prelude.foldr consM nil
...
 
--- | Keep reading 'Storable' elements from 'Ptr' onwards.
+-- | Keep reading 'Storable' elements from an immutable 'Ptr' onwards.
 --
 -- /Unsafe:/ The caller is responsible for safe addressing.
 --
 -- /Pre-release/
 {-# INLINE fromPtr #-}
-fromPtr :: forall m a. (MonadIO m, Storable a) => Ptr a -> Stream m a
+fromPtr :: forall m a. (Monad m, Storable a) => Ptr a -> Stream m a
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Stream/StreamD/Generate.hs:1153]: Why make this change? What is the use-case for this?

Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Stream/StreamD/Generate.hs:1153]: I have an eerie feeling about it. Can't explain it.

**Review by adithyaov — APPROVED** [review]

**Comment by adithyaov** [comment]
Some comments.

File: core/src/Streamly/Internal/Data/Array/Type.hs (hunk 4 above)
Comment (harendra-kumar) [reply core/src/Streamly/Internal/Data/Array/Type.hs:525]: not exposed, not meant to be exposed. Experimental and to be used internally where it is safe.

**Pull request merged**